### ISSUE TITLE
Filenames with colons

### ISSUE THREAD

**Issue by dasisdormax** [redox-os/redox#903]
As the colon separates the scheme and the reference (filename), things may behave weird or unexpected when working with files that include it in their filename.

While (i guess) not many programs need to work with filenames with colons, they should be able to work without having to write different code for Redox/non-Redox operating systems.

### Examples

**Example 1**:
- Boot into Redox, open a shell as user and execute
~~~
touch file:/home/user/:
ls
~~~
- ls will fail (trying to get information about the root of the unnamed scheme) outputting just:
~~~
error: permission denied
~~~

**Example 2**:
- Boot into Redox, open a shell as user and execute
~~~
touch file:/home/user/file:
ls -l
~~~
- ls will output the file information about the root directory of the file scheme instead of the file "file:" within the current working directory:
~~~
-rw-------  1000  1000 [...] LICENSE
-rw-------  1000  1000 [...] README.md
drwxr-xr-x     0     0 [...] file:
~~~

**Build information (for both examples)**:
self-built redox
rustc 1.17.0-nightly (3da40237e 2017-03-24)
git revision 24f080b4fca8555b15de89272697dfb8062226dd (March 25)

### What can we do?

**Currently working:**
For relative filenames, you can prepend the current working directory to everything. If you use `ls $PWD` in the examples above, everything works as expected as $PWD already contains the scheme.

**Suggestion, not working today:**
For paths including slashes (such as the absolute "/home/user/:" or the relative "./:"), it is clear (at least to me) that the program wants to access that file, not a scheme called "/home/user/" or "./". These variants would mean the same on both Redox and non-Redox operating systems.

This does assume though that scheme names can not contain slashes. If it is intended to support such scheme names, this would obviously not be possible.

### Discuss!

**Comment by DaanHoogland** [comment]
@dasisdormax good point and also discussed in https://github.com/redox-os/coreutils/issues/102

**Comment by DaanHoogland** [comment]
@dasisdormax good point and also discussed in https://github.com/redox-os/coreutils/issues/102

**Comment by bungcip** [comment]
why not just forbid it? On windows you cannot use it in filename. NTFS forbid `0x00-0x1F 0x7F * / : < > ? \ |` in filename. I think such restrictions is reasonable. Sometimes having too much flexibility in filename (linux allow anything as along as not `NUL /`) can lead to security problems  (https://lwn.net/Articles/686789/)

**Comment by bungcip** [comment]
another good reading: https://www.dwheeler.com/essays/fixing-unix-linux-filenames.html

**Comment by dasisdormax** [comment]
Nice input! This would definitely make things easier for shell developers. I think we all agree that control characters in filenames are a bad idea.

For characters as `* : ?` however, we should only prohibit the creation of these files, not reading. One would at least need a way to access or rescue an existing 'bad' file on a USB or Network drive.

So I rather want see this **in addition** to fixing access to colon files. Anyway, this is a good topic to discuss!

**Comment by ritschwumm** [comment]
as a programmer i have to say that i very much prefer the way unix treats filenames.
having only two rules (slash separates, null terminates) makes it much easier to write correct code than when you have to remember half a dozen _different_ rule sets, depending on OS and file system. i've written some cross-platform code, and having to deal with all those rules (and not forgetting one - most notably windows with its weird special folder names, and HFS+ with its unicode "normalization") is a PITA. keep it simple...

imho the problem is not that paths are restricted enough, the (worst) problem is that shells are doing way too much magic _on top_ of that that again nobody is able to remember. adding more rules does _not_ improve the situation - on the contrary, it would make it even more difficult to get everything right.

**Comment by ametisf** [comment]
From point of view of regular user, being prohibited to use some character in file names puts me off. Having as little restrictions as possible (NULL and /) is IMO more user friendly.

**Comment by ksf** [comment]
The way Unix treats filenames, having only two rules, is already restrictive enough to need an user-defined escaping layer to store arbitrary data: If you want to store control characters you probably also want to store `\0` and `/`. 

That utter freedom also clashes with the usual "things are text" principle, as any such usage is not user-readable. It also already messes up practically *all* software out there that does any amount of path parsing as they just don't assume such use-cases. 

That is, it's broken as a binary store, it's broken as a user-interactible store.

Should redox provide (hierarchical, persistent) k/v stores which can use arbitrary  binary keys? I'm inclined to say yes. I'd also say that `file:` probably isn't the right place for it.

Something like "starts with a printable unicode glyph or `.`, plus a handful of symbols and literal space" would work just fine, the rest can use escapes. Maybe even require DNS-like normalisation, that is, don't allow e.g. Latin and Cyrillic to be mixed without intervening symbols and such. You could then `rm file:/home/foo/%3Aw` if you messed up your vi-foo.

It should also be noted that this doesn't need to apply to all schemes: It might very well make sense to address a thing such as (hypothetical) `kv:` in binary, at least optionally. That shouldn't be exposed to standard shell and file management utilities, though. And we probably should have a standard encoding scheme to use `open()` with length-tagged binary data.

**Comment by le-jzr** [comment]
The problem with filename characters basically boils down to unsanitized script input. Without limitations, every filename is a potential code injection, and UNIX shells traditionally ignore that issue. While prohibiting the characters is a simple solution that mostly works, what you really want is a shell language that does escaping transparently. Of course, that means being incompatible with most existing shells.

**Comment by DemiMarie** [comment]
@le-jzr I agree.  That also boils down to most existing shell languages being weakly typed.

I think that Microsoft’s Powershell actually gets this close to correct.  What is necessary is a shell that is *strongly* typed, and does not do any automatic conversion.  Except possibly between integers and strings.

**Comment by le-jzr** [comment]
But even knowing that, any attempt at building a well-structured shell language will fail due to inertia. If it is possible to use Bourne-compatible scripts, people will keep using them, because that's what they have known for decades (not to mention the bulk of readily available footguns on stackoverflow etc).

It might sound brutal, but imo the best solution would be for the APIs to work exclusively with mangled paths using something similar to IDNA, so that even a straight port of bash will be safe by default, and leave it to applications to de/mangle names on UI boundary.

**Comment by logannc** [comment]
> For characters as * : ? however, we should only prohibit the creation of these files, not reading. One would at least need a way to access or rescue an existing 'bad' file on a USB or Network drive.

At first glance, this isn't a terrible idea, but it actually is. Imagine you are required, by some external process you cannot change, to upload or create files with a timestamp in the name (or of a directory, etc). This is an extremely common format on linux due to it's permissiveness. Being _unable_ to do so on a platform would be unacceptable.

Prohibiting `NUL` and `/` are plenty sufficient rules. The difficulty in using paths that include spaces or unusual characters comes from the input layer parsing them incorrectly - bash interpreting the space as a separator, colons in URIs, etc. The filesystem doesn't need to worry about that. Those are purely UX concerns.

**Comment by le-jzr** [comment]
Perhaps the limitations should be strict by default, but allow individual processes (or directories) to turn them off. That would take care about cases where user absolutely wants those names, at their own responsibility, but also be safe by default.

If you think it's not your/our problem if a random sysadmin uses badly written scripts, think again. If you allow people to do things that are vulnerable-by-default, they will make them vulnerable. It's your data/passwords/money that's on the line, unless you decide never to use any online services.

**Comment by DemiMarie** [comment]
Newlines should definitely be forbidden.

Spaces should be allowed in the GUI, but be encoded in the actual APIs.  I
think that URL-encoding might be a good option (though please do not
require escaping non-ASCII characters).

**Comment by jackpot51** [comment]
The more I think about it, the more I think we should either disallow colons completely after the scheme, or leave it up to the scheme to interpret them. I would not foresee any other special characters than `:`. Rules more complicated than that are not reasonable to have at a system-wide level.

Another alternative is to use another character for the scheme seperator, but I think it would affect readability. Very few file names in Linux use `:`.

**Comment by DemiMarie** [comment]
Can we require that scheme names contain only certain characters?
